checker/src/org/checkerframework/checker/upperbound/UpperBoundVisitor.java

```diff
@@ -0,0 +1,131 @@
+     * checks if the index has been assigned a reasonable UpperBound type: only an index with type
+     * LTLengthOf(arr) is safe to access arr. If that fails, it checks if the access is still safe.
+     * To do so, it checks if the MinLen checker knows the minimum length of arr by querying the
+     * MinLenATF. If the MinLen of the array is known, the visitor can check if the index is less
+     * than the MinLen, using the Value Checker. If so then the access is still safe. Otherwise,
+     * report a potential unsafe access.
+     */
+    @Override
+    public Void visitArrayAccess(ArrayAccessTree tree, Void type) {
+        ExpressionTree indexTree = tree.getIndex();
+        ExpressionTree arrTree = tree.getExpression();
+        String arrName = FlowExpressions.internalReprOf(this.atypeFactory, arrTree).toString();
+        AnnotatedTypeMirror indexType = atypeFactory.getAnnotatedType(indexTree);
+
+        // Need to be able to check these as part of the conditional below.
+        // Find max because it's important to determine whether the index is
+        // less than the minimum length of the array. If it could be any
+        // of several values, only the max is of interest.
+        Integer valMax = atypeFactory.valMaxFromExpressionTree(indexTree);
+        Integer minLen = atypeFactory.minLenFromExpressionTree(arrTree);
+
+        // Is indexType LTL/LTOM of a set containing arrName?
+        if ((indexType.hasAnnotation(LTLengthOf.class)
+                        || indexType.hasAnnotation(LTOMLengthOf.class))
+                && (UpperBoundUtils.hasValue(indexType, arrName)
+                        || UpperBoundUtils.hasValue(indexType, arrTree.toString()))) {
+            // If so, this is safe - get out of here.
+            return super.visitArrayAccess(tree, type);
+        } else if (valMax != null && minLen != null && valMax < minLen) {
+            return super.visitArrayAccess(tree, type);
+        } else {
+            // Unsafe, since neither the Upper bound or MinLen checks succeeded.
+            checker.report(Result.warning(UPPER_BOUND, indexType.toString(), arrName), indexTree);
+            return super.visitArrayAccess(tree, type);
+        }
+    }
+
+    @Override
+    public Void visitMethodInvocation(MethodInvocationTree tree, Void type) {
+        if (isFirstArgListMethod(tree)) {
+            ExpressionTree indexTree = tree.getArguments().get(0);
+            ExpressionTree lstTree = tree.getMethodSelect();
+            String lstName = FlowExpressions.internalReprOf(this.atypeFactory, lstTree).toString();
+            AnnotatedTypeMirror indexType = atypeFactory.getAnnotatedType(indexTree);
+            String[] parts = lstTree.toString().split("\\.");
+            // the last part of the methodName is the method (get, remove, etc) so get second to last part
+            String localName = parts[parts.length - 2];
```

Comment:
`localName` should not be needed.  `lstName` should be sufficient.  If it's not, then there is a problem with expression annotation support that needs to be fixed.